---
title: "Nmen_microbiome_2020_2021.ALL"
author: "Yair Motro"
date: "15/02/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Analysis of second sequencing run and repeats

```{r load-libs, include=FALSE}
library(tidyverse)
library(qiime2R)
library(microeco)
library(agricolae)
library(randomForest)
library(ggpubr)
library(patchwork)
library(revealjs)
library(picante)
library(splitstackshape)
library(rstatix)
library(ggtree)
library(dplyr)
library(rtk)
library(ggExtra)
```


## Note:

- ns: p > 0.05
- *: p <= 0.05
- **: p <= 0.01
- ***: p <= 0.001
- ****: p <= 0.0001

```{r load-ALL-combined-data, include=FALSE}

rerun_infile='table-dada2.qza'
rerun_metadata<-read_q2metadata("00.metadata.combined.2020_2021.tsv")
rerun_taxonomy<-read_qza("taxonomy-dada2.qza")$data %>% parse_taxonomy()
rerun_SVs<-read_qza(rerun_infile)$data
rerun_phylo_tree <- read_qza('rooted-tree-dada2.qza')$data

#rerun_metadata_H <- filter(rerun_metadata, type == 'H')
#rerun_metadata_R <- filter(rerun_metadata, type == 'R')
rerun_metadata_only_3repeats_all<-filter(rerun_metadata, pid == '1014' | pid == '1029' | pid == '1031' | pid == '1040' | pid == '1057' | pid == '1173' | pid == '1351' | pid == '1486' | pid == '1544' | pid == '1719' | pid == '1784' | pid == '1848' | pid == '2457' | pid == '2488')
rerun_metadata_only_3repeats<-filter(rerun_metadata_only_3repeats_all, SampleID != "MG049-1057-2" & SampleID != "622-Jacob005-Re-1848-0" & SampleID != "622-Jacob004-Re1784-0" & SampleID != "622-Jacob007-Re2488-0" & SampleID != "MG047-1173-2" & SampleID != "MG042-1544-2" & SampleID != "MG045-1351-2" & SampleID != "MG043-1486-2")

#rerun_metadata_H_only_repeats %>% arrange(animal)
#rerun_metadata_G0 <- filter(rerun_metadata, study_group == '0')
#rerun_metadata_G1 <- filter(rerun_metadata, study_group == '1')
#rerun_metadata
#rerun_metadata_only_3repeats
# set.seed is used to fix the random number generation to make the results repeatable
set.seed(123)
# make the plotting background same with the tutorial
theme_set(theme_bw())
class(rerun_SVs)
class(rerun_taxonomy)
class(rerun_metadata)
class(rerun_metadata_only_3repeats)
#class(rerun_metadata_H)
#class(rerun_metadata_R)
#class(rerun_metadata_H_only_repeats)
#class(rerun_metadata_G1)
#metadata <- data.frame(metadata, row.names = 1)
#keep column
rownames(rerun_metadata) <- rerun_metadata[,1]
rownames(rerun_metadata_only_3repeats) <- rerun_metadata_only_3repeats[,1]
#rownames(rerun_metadata_H) <- rerun_metadata_H[,1]
#rownames(rerun_metadata_R) <- rerun_metadata_R[,1]
#rownames(rerun_metadata_H_only_repeats) <- rerun_metadata_H_only_repeats[,1]
#rownames(rerun_metadata_G1) <- rerun_metadata_G1[,1]
#rownames(metadata_both_only) <- metadata_both_only[,1]
# In R6 class, '$new' is the original method used to create a new object of class
rerun_dataset <- microtable$new(sample_table = rerun_metadata, otu_table = rerun_SVs, tax_table = rerun_taxonomy, phylo_tree = rerun_phylo_tree)
rerun_dataset_triple <- microtable$new(sample_table = rerun_metadata_only_3repeats, otu_table = rerun_SVs, tax_table = rerun_taxonomy, phylo_tree = rerun_phylo_tree)

#rerun_dataset_H <- microtable$new(sample_table = rerun_metadata_H, otu_table = rerun_SVs, tax_table = rerun_taxonomy, phylo_tree = rerun_phylo_tree)
#rerun_dataset_R <- microtable$new(sample_table = rerun_metadata_R, otu_table = rerun_SVs, tax_table = rerun_taxonomy, phylo_tree = rerun_phylo_tree)
#rerun_dataset_H_only_repeats <- microtable$new(sample_table = rerun_metadata_H_only_repeats, otu_table = rerun_SVs, tax_table = rerun_taxonomy, phylo_tree = rerun_phylo_tree)

class(rerun_dataset)
rerun_dataset$tidy_dataset()
#print(rerun_dataset)
rerun_dataset$tidy_dataset()
rerun_dataset$tax_table %<>% base::subset(Kingdom == "Archaea" | Kingdom == "Bacteria")
class(rerun_dataset)
rerun_dataset$tidy_dataset()
#print(rerun_dataset)
# This will remove the lines containing the taxa word regardless of taxonomic ranks and ignoring word case in the tax_table.
# So if you want to filter some taxa not considerd pollutions, please use subset like the previous operation.
rerun_dataset$filter_pollution(taxa = c("mitochondria", "chloroplast"))
rerun_dataset$tidy_dataset()
rerun_dataset$sample_sums() %>% sort() %>% median
#rerun_dataset$sample_sums() %>% sort()
#selected 6900 - from qiime2 table-dada2.qzv looks to keep all samples
rerun_dataset$rarefy_samples(5000)
rerun_dataset$tidy_dataset()
rerun_dataset$cal_abund()
class(rerun_dataset$taxa_abund)
class(rerun_dataset)


class(rerun_dataset_triple)
rerun_dataset_triple$tidy_dataset()
#print(rerun_dataset_triple)
rerun_dataset_triple$tidy_dataset()
rerun_dataset_triple$tax_table %<>% base::subset(Kingdom == "Archaea" | Kingdom == "Bacteria")
class(rerun_dataset_triple)
rerun_dataset_triple$tidy_dataset()
#print(rerun_dataset_triple)
# This will remove the lines containing the taxa word regardless of taxonomic ranks and ignoring word case in the tax_table.
# So if you want to filter some taxa not considerd pollutions, please use subset like the previous operation.
rerun_dataset_triple$filter_pollution(taxa = c("mitochondria", "chloroplast"))
rerun_dataset_triple$tidy_dataset()
rerun_dataset_triple$sample_sums() %>% sort() %>% median
rerun_dataset_triple$sample_sums() %>% sort()
#selected 5800 - to include all triplicate samples
rerun_dataset_triple$rarefy_samples(5000)
rerun_dataset_triple$tidy_dataset()
rerun_dataset_triple$cal_abund()
class(rerun_dataset_triple$taxa_abund)
class(rerun_dataset_triple)


```

## Run1 rarefied @ 8000 stats

```{r print-rerun}

print(rerun_dataset)
print(rerun_dataset_triple)

```

```{r phyla-genera-abundances-ALL}

rerun_phyla <- trans_abund$new(dataset = rerun_dataset, taxrank = "Phylum")#, groupmean = "run")
rerun_genera <- trans_abund$new(dataset = rerun_dataset, taxrank = "Genus")
rerun_dataset$cal_betadiv(unifrac = TRUE)

rerun_phyla$plot_bar()+theme(axis.text.x = element_text(size = 7, angle = 90),axis.text.y = element_text(size = 7))
rerun_genera$plot_bar()+theme(axis.text.x = element_text(size = 7, angle = 90),axis.text.y = element_text(size = 7))


# return dataset$beta_diversity
class(rerun_dataset$beta_diversity)

# we first create an object and select PCoA for ordination
rerun <- trans_beta$new(dataset = rerun_dataset, group = "set", measure = "bray", ordination = "PCoA")

#rerun_t1$res_ordination is the ordination result list
class(rerun$res_ordination)
rerun$res_ordination
# plot the PCoA result
p1a<-rerun$plot_ordination(plot_color = "year", plot_shape = "time", plot_group_ellipse = TRUE)+ labs(title = "Nmen2020.2021.ALL - ALL seq sets",
        #subtitle = "pt2 - batches",
        caption = "all samples") + theme(
    axis.title.y=element_text(size=10,face="bold"),
    axis.title.x=element_text(size=10,face="bold"),
    axis.text.y=element_text(size=8),
    axis.text.x=element_text(size=8),
    plot.title = element_text(hjust = 0.5, size = 14),    # Center title position and size
    plot.subtitle = element_text(hjust = 0.5),            # Center subtitle
    plot.caption = element_text(hjust = 0.5, face = "italic")# move caption to the left
  )
ggExtra::ggMarginal(p1a,x=batch,y=batch,type="density", groupFill=TRUE, groupColour = TRUE)

# we first create an object and select PCoA for ordination
rerun.samp_time <- trans_beta$new(dataset = rerun_dataset, group = "time", measure = "bray", ordination = "PCoA")

#rerun_t1$res_ordination is the ordination result list
class(rerun.samp_time$res_ordination)

# plot the PCoA result
p1b<-rerun.samp_time$plot_ordination(plot_color = "time", plot_shape = "year", plot_group_ellipse = TRUE)+ labs(title = "Nmen2020.2021.ALL - all seq sets") + theme(
    axis.title.y=element_text(size=10,face="bold"),
    axis.title.x=element_text(size=10,face="bold"),
    axis.text.y=element_text(size=8),
    axis.text.x=element_text(size=8),
    plot.title = element_text(hjust = 0.5, size = 14),    # Center title position and size
    plot.subtitle = element_text(hjust = 0.5),            # Center subtitle
    plot.caption = element_text(hjust = 0.5, face = "italic")# move caption to the left
  )
ggExtra::ggMarginal(p1b,x=batch,y=batch,type="density", groupFill=TRUE, groupColour = TRUE)

rerun.samp_time$cal_group_distance(within_group = TRUE)
rerun.samp_time$plot_group_distance(distance_pair_stat = TRUE, pair_compare_method = "t.test")+ggtitle("Nmen2020.2021.ALL - all samples from all sets")#+ggsave("combined.20210202/rarefied_5000.all_times.BrayCurtis.group_sig.png")

```

```{r phyla-genera-abundances-ALL-triplicates}

rerun_dataset_triple$sample_table$pid%>%sort()
rerun_phyla_triple <- trans_abund$new(dataset = rerun_dataset_triple, taxrank = "Phylum")#, groupmean = "run")
rerun_genera_triple <- trans_abund$new(dataset = rerun_dataset_triple, taxrank = "Genus")
rerun_dataset_triple$cal_betadiv(unifrac = TRUE)

rerun_phyla_triple$plot_bar()+theme(axis.text.x = element_text(size = 7, angle = 90),axis.text.y = element_text(size = 7))
rerun_genera_triple$plot_bar()+theme(axis.text.x = element_text(size = 7, angle = 90),axis.text.y = element_text(size = 7))


# return dataset$beta_diversity
class(rerun_dataset_triple$beta_diversity)

# we first create an object and select PCoA for ordination
rerun_triple <- trans_beta$new(dataset = rerun_dataset_triple, group = "set", measure = "bray", ordination = "PCoA")

#rerun_t1$res_ordination is the ordination result list
class(rerun_triple$res_ordination)

rerun_triple$sample_table %>% arrange(by=pid)
# plot the PCoA result
p1_triple<-rerun_triple$plot_ordination(plot_color = "year", plot_shape = "pid", plot_group_ellipse = TRUE)+ labs(title = "Nmen2020.2021.ALL - all duplicated seq runs") + theme(
    axis.title.y=element_text(size=10,face="bold"),
    axis.title.x=element_text(size=10,face="bold"),
    axis.text.y=element_text(size=8),
    axis.text.x=element_text(size=8),
    plot.title = element_text(hjust = 0.5, size = 14),    # Center title position and size
    plot.subtitle = element_text(hjust = 0.5),            # Center subtitle
    plot.caption = element_text(hjust = 0.5, face = "italic")# move caption to the left
  )
ggExtra::ggMarginal(p1_triple,x=batch,y=batch,type="density", groupFill=TRUE, groupColour = TRUE)

rerun_triple$cal_group_distance(within_group = TRUE)
rerun_triple$plot_group_distance(distance_pair_stat = TRUE, pair_compare_method = "t.test")+ggtitle("Nmen2020.2021.ALL - samples duplicated in runs")#+ggsave("combined.20210202/rarefied_5000.duplicates_runs.BrayCurtis.group_sig.png")
```

```{r}
library("xtable")
library("cluster")
library("ruv")
library("lmerTest")
library("bapred")
library("mixOmics")
library("limma")
library("AgiMicroRna")
library("variancePartition")
library("pvca")
library("sva")
library("microbiome")
library("phyloseq")
library("microeco")
library("ggExtra")
source("Functions.R")

rerun_dataset_for_limma <- microtable$new(sample_table = rerun_metadata, otu_table = rerun_SVs, tax_table = rerun_taxonomy, phylo_tree = rerun_phylo_tree)
class(rerun_dataset_for_limma)
rerun_dataset_for_limma$tidy_dataset()
#print(rerun_dataset)
rerun_dataset_for_limma$tax_table %<>% base::subset(Kingdom == "Archaea" | Kingdom == "Bacteria")
class(rerun_dataset_for_limma)
rerun_dataset_for_limma$tidy_dataset()
#print(rerun_dataset)
# This will remove the lines containing the taxa word regardless of taxonomic ranks and ignoring word case in the tax_table.
# So if you want to filter some taxa not considerd pollutions, please use subset like the previous operation.
rerun_dataset_for_limma$filter_pollution(taxa = c("mitochondria", "chloroplast"))
rerun_dataset_for_limma$tidy_dataset()
rerun_dataset_for_limma$sample_sums() %>% sort() %>% median

test<-rerun_dataset_for_limma$otu_table
test.meta<-rerun_dataset_for_limma$sample_table
test.meta.year<-data.frame(ID=test.meta[,1],year=test.meta[,4])
test.meta.time<-data.frame(ID=test.meta[,1],time=test.meta[,3])
test.meta.year
test.meta.time
test <- test + 1
test.clr <- logratio.transfo(test, logratio = 'CLR')
class(test.clr) <- 'matrix'
test.clr
test.meta.year.matrix<-as.matrix(test.meta.year$year)
class(test.meta.year.matrix)<-'numeric'
test.meta.time.matrix<-as.matrix(test.meta.time$time)
class(test.meta.time.matrix)<-'numeric'

dataset.limma <- removeBatchEffect(test.clr, batch = test.meta.year.matrix, design = test.meta.time.matrix)
dataset.limma

dataset.pca<- pca(test.clr, ncomp = 3)
dataset.pca.limma <- pca(dataset.limma, ncomp = 3)

physeq <- meco2phyloseq(rerun_dataset_for_limma)
physeq

pre.batcheff.year.plot<-plot_ordination(physeq, dataset.pca, color = "year") +
                geom_point(size = 5)+ggtitle("pre-correction")
post.batcheff.year.plot<-plot_ordination(physeq, dataset.pca.limma, color = "year") +
                geom_point(size = 5)+ggtitle("post-correction")

pre.batcheff.time.plot<-plot_ordination(physeq, dataset.pca, color = "time") +
                geom_point(size = 5)+ggtitle("pre-correction")
post.batcheff.time.plot<-plot_ordination(physeq, dataset.pca.limma, color = "time") +
                geom_point(size = 5)+ggtitle("post-correction")

gridExtra::grid.arrange(pre.batcheff.year.plot,post.batcheff.year.plot)
gridExtra::grid.arrange(pre.batcheff.time.plot,post.batcheff.time.plot)
rerun_dataset_for_limma$sample_sums()%>%sort()


``` 
